File: src/test/java/community/whatever/onembackendjava/api/controller/UrlShortenControllerTest.java
package community.whatever.onembackendjava.api.controller;

import com.fasterxml.jackson.core.type.TypeReference;
import community.whatever.onembackendjava.api.dto.ShortenUrlDto;
import community.whatever.onembackendjava.setting.RestDocsTestSupport;
import org.junit.jupiter.api.Test;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.restdocs.payload.JsonFieldType;
import org.springframework.test.web.servlet.ResultActions;
import org.springframework.util.LinkedMultiValueMap;
import org.springframework.util.MultiValueMap;

import java.util.HashMap;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;
import static org.springframework.restdocs.headers.HeaderDocumentation.headerWithName;
import static org.springframework.restdocs.headers.HeaderDocumentation.requestHeaders;
import static org.springframework.restdocs.mockmvc.RestDocumentationRequestBuilders.get;
import static org.springframework.restdocs.mockmvc.RestDocumentationRequestBuilders.post;
import static org.springframework.restdocs.payload.PayloadDocumentation.*;
import static org.springframework.restdocs.request.RequestDocumentation.parameterWithName;
import static org.springframework.restdocs.request.RequestDocumentation.queryParameters;
import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

class UrlShortenControllerTest  extends RestDocsTestSupport {

    @Test
    void shortenUrlSearch() throws Exception{
        ShortenUrlDto.Get.Request dto = new ShortenUrlDto.Get.Request();
        dto.setKey("11111111");

        MultiValueMap<String, String> params = new LinkedMultiValueMap<>();
        Map<String, String> map = objectMapper.convertValue(dto, new TypeReference<Map<String, String>>() {
        });
        params.setAll(map);

        ResultActions result = this.mockMvc.perform(
                get("/shorten-url")
                        .contentType(contentType)
                        .accept(contentType)
                        .params(params)
        );

        result.andExpect(status().isOk())
                .andDo(print())
                .andDo(restDocs.document(
                                queryParameters(
                                        parameterWithName("key").description("short url key")
                                ),
                                responseFields(
                                        fieldWithPath("resultCode").type(JsonFieldType.STRING).description("결과 코드"),
                                        fieldWithPath("msg").type(JsonFieldType.STRING).description("결과 메시지"),
                                        fieldWithPath("timeZone").type(JsonFieldType.STRING).description("서버 타임존"),
                                        fieldWithPath("timeStamp").type(JsonFieldType.NUMBER).description("서버 응답시간"),
                                        fieldWithPath("data").type(JsonFieldType.OBJECT).description("결과 객체").optional(),
                                        fieldWithPath("data.originUrl").type(JsonFieldType.STRING).description("원본 url").optional()
                                )
                        )
                );
    }

    @Test
    void shortenUrlCreate() throws Exception {
        ShortenUrlDto.Create.Request dto  = new ShortenUrlDto.Create.Request();
        dto.setOriginUrl("https://www.google.com");

        ResultActions result = this.mockMvc.perform(
                post("/shorten-url")
                        .contentType(contentType)
                        .accept(contentType)
                        .content(objectMapper.writeValueAsString(dto))
        );

        result.andExpect(status().isOk())
                .andDo(print())
                .andDo(restDocs.document(
                                requestFields(
                                        fieldWithPath("originUrl").type(JsonFieldType.STRING).description("요청 url")
                                ),
                                responseFields(
                                        fieldWithPath("resultCode").type(JsonFieldType.STRING).description("결과 코드"),
                                        fieldWithPath("msg").type(JsonFieldType.STRING).description("결과 메시지"),
                                        fieldWithPath("timeZone").type(JsonFieldType.STRING).description("서버 타임존"),
                                        fieldWithPath("timeStamp").type(JsonFieldType.NUMBER).description("서버 응답시간"),
                                        fieldWithPath("data").type(JsonFieldType.OBJECT).description("결과 객체").optional(),
                                        fieldWithPath("data.key").type(JsonFieldType.STRING).description("생성된 키").optional()
                                )
                        )
                );
    }
}

File: src/test/kotlin/community/whatever/onembackendkotlin/presentation/UrlShortenControllerTest.kt
package community.whatever.onembackendkotlin.presentation

import com.fasterxml.jackson.databind.ObjectMapper
import community.whatever.onembackendkotlin.application.UrlShortenService
import community.whatever.onembackendkotlin.application.exception.UrlNotFoundException
import community.whatever.onembackendkotlin.domain.ShortenedUrl
import community.whatever.onembackendkotlin.presentation.dto.OriginUrlResponse
import community.whatever.onembackendkotlin.presentation.dto.ShortenUrlCreateRequest
import community.whatever.onembackendkotlin.presentation.dto.ShortenUrlSearchRequest
import community.whatever.onembackendkotlin.presentation.dto.ShortenedUrlResponse
import net.datafaker.Faker
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.mockito.BDDMockito.given
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest
import org.springframework.http.MediaType
import org.springframework.test.context.bean.override.mockito.MockitoBean
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
import java.time.LocalDateTime

@WebMvcTest(UrlShortenController::class)
class UrlShortenControllerTest(
    @Autowired private val mockMvc: MockMvc,
) {

    @MockitoBean
    private lateinit var urlShortenService: UrlShortenService

    private val objectMapper = ObjectMapper()
    private val faker = Faker()

    @Test
    fun `원본 URL 등록 시 키를 반환한다`() {
        // given
        val originUrl = faker.internet().url()
        val shortenedUrl = ShortenedUrl(faker.idNumber().peselNumber(), originUrl, LocalDateTime.now())
        val request = ShortenUrlCreateRequest(originUrl)
        given(urlShortenService.saveShortenUrl(originUrl)).willReturn(shortenedUrl)

        // when
        val result = mockMvc.perform(
            post("/shorten-url/create")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(request))
        )
            .andExpect(status().isOk)
            .andReturn()
            .response
            .contentAsString

        // then
        assertThat(result).isNotBlank()
    }

    @Test
    fun `키로 검색하면 원본 URL을 반환한다`() {
        // given
        val shortenedUrl = faker.idNumber().peselNumber()
        val request = ShortenUrlSearchRequest(shortenedUrl)
        val originUrl = faker.internet().url()
        val response = OriginUrlResponse(originUrl)
        given(urlShortenService.getOriginUrl(shortenedUrl)).willReturn(originUrl)

        // when
        val result = mockMvc.perform(
            post("/shorten-url/search")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(request))
        )
            .andExpect(status().isOk)
            .andReturn()
            .response
            .contentAsString

        // then
        assertThat(result).isEqualTo(objectMapper.writeValueAsString(response))
    }

    @Test
    fun `유요하지 않은 키로 검색하면 404 에러를 반환한다`() {
        // given
        val notExistShortenedUrl = faker.idNumber().peselNumber()
        val request = ShortenUrlSearchRequest(notExistShortenedUrl)
        given(urlShortenService.getOriginUrl(notExistShortenedUrl)).willThrow(UrlNotFoundException())

        // when
        mockMvc.perform(
            post("/shorten-url/search")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(request))
        )
            .andExpect(status().isNotFound)
    }

    @Test
    fun `중복된 원본 URL을 등록하면 기존 키를 반환한다`() {
        // given
        val originUrl = "https://www.google.com"
        val request = ShortenUrlCreateRequest(originUrl)
        val shortenedUrl = ShortenedUrl(faker.idNumber().peselNumber(), originUrl, LocalDateTime.now())
        given(urlShortenService.saveShortenUrl(originUrl)).willReturn(shortenedUrl)

        // when
        val result = mockMvc.perform(
            post("/shorten-url/create")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(request))
        )
            .andExpect(status().isOk)
            .andReturn()
            .response
            .contentAsString

        // then
        assertThat(result).isEqualTo(objectMapper.writeValueAsString(ShortenedUrlResponse(shortenedUrl.id)))
    }
}
